Unity-Technologies/ScriptableRenderPipeline
# Override ShaderGUI per ShaderGraph [skip ci]

**Summary:**

- Back end changes such that the entire generated shaders now only contains a single CustomEditor "ShaderGUI" line, as what we did prior (per SubShader) was not supported by Unity.
- Added a Toggle & TextField to each Master Node allowing the user to override the ShaderGUI per Shader Graph:
![ShaderGUIOverridePresentation0](https://user-images.githubusercontent.com/3393041/76673926-5a586200-6567-11ea-8110-2bb8d5241a3b.gif)
Toggle to open up the Text Field, which automatically gets populated with the SRP's default, and if an invalid ShaderGUI is given an warning badge is thrown.

While a lot of simple changes in the Master Nodes, a lot is needed changes for the future of ShaderGraphShaderGUI and will carry over into Stacks and the Internal Inspector.

---
**Implementation Notes & Notes to Reviewers:**

- I would say this copy & pasted code in each master node would belongs in a shared MasterNode class, but since Targets removed the old MasterNode class as a step towards the master node-less future I did what I did. That is to say all MasterNodes now extend `ICanChangeShaderGUI `and have the same code starting with `[SerializeField] private string m_ShaderGUIOverride` and ending with `set => m_OverrideEnabled = value`. This is except for the Sprite MasterNodes, which now have a shared super simple settings plane class.
- First tooltip in a Node Settings pane used here. This should be easily transferable into the Internal Inspector.
- ITargetImplementation now has a new method for getting the sub descriptors depending on the master node, was needed.
- `MasterNodeSettingsView` has the front end changes.

---
**Notes to QA:**

- _Known Issue:_ If a Master Node has a Red Error Badge (for example, no SRP assigned), and then you change the ShaderGUI to something incorrect (such that a White Warning Badge spawns), then fix the issue, the Red Error doesn't reappear. Ideally the White Warning Badge doesn't override the Error anyways. I am look into this, but so far it seems like a wider bug with the MessageManager.
- If you don't know what ShaderGUI is, this is useful: https://docs.unity3d.com/Manual/SL-CustomShaderGUI.html, recommended to at least get surface level familiar with it as more ShaderGUI will be coming in the future! Basically a single line is inserted inside the generated shader that determines which ShaderGUI is used when selecting a Material that uses the Shader Graph.
- We are only inserting a single line inside the generated shader, as such the complexity of the ShaderGUI used is irrelevant to these changes.
- If you do not want to write your own custom ShaderGUI, I can give you some really crappy ShaderGUIs I tested with during development. Again, nothing fancy needed for the testing of this specific feature.
- In order to use the default trunk ShaderGUI, type in BaseShaderGUI into the field.
- Remember that HDRP & URP use different ShaderGUIs for the same Master Node, which will come up in testing. For exmample, if you set Override ShaderGUI, does it first display the pipeline's default?
- You need to save the SG for the material to use the new ShaderGUI.

---
**Manually Tested:**

- Undo & Redo all new actions (i.e. changing of the new fields).
- Type in all sorts of junk into the new text field, confirm only allowed characters allowed (ones that could be in a C# class).
- Closing and reopening Unity and confirming that these two fields are in still in the same state.
- Bigger shaders.
- Confirmed that the master nodes and the ShaderGUI textfield (once Override ShaderGUI is checked) spawn with the right ShaderGUI AFTER switching between different SRPs.
- That every single Master Node still spawns with the correct ShaderGUI with OverrideEnabled is unchecked.
- Checking the generated code after switching active masters and undo & redo operations; confirming that the CustomEditor override is what it should be.
- Confirmed that there is only one CustomEditor line ever regardless of the # of implementations (current # of different SRPs to choose from).
- Confirmed that Master Nodes that aren't compatible with the current SRP don't spawn the CustomEditor line in the generated shader, in addition the ShaderGUI TextField doesn't have any text when Override ShaderGUI is checked.
- Confirmed that this works with custom user namespaces, such as Testing.Bull.doubleTime, ect.

---
**Technical Risk:** 3/4
**Halo Effect:** 1/4, affects nothing else, aside for a minor change in the generator.

---
**Notes to Documentation:**
https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/6192/commits/f1cbef8b20ced2ea6dbbde903ace8d845333cc22

All Master Nodes with documentation have this:

> | **Override ShaderGUI**        | Enable overriding of the [https://docs.unity3d.com/ScriptReference/ShaderGUI.html](ShaderGUI) that this Shader Graph uses, spawning the "ShaderGUI" field below. |
> | **ShaderGUI**                 | The full name of the ShaderGUI class (including the class path) to use. A ShaderGUI class inside the UnityEditor namespace may omit UnityEditor. |

Meanwhile HD will have this documentation under Material Inspector:

>  These section does not apply if **Override ShaderGUI** is enabled.

---
**Yamato:**
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/sg%252Fshadergui-override-per-sg
